[2024-08-26T10:47:15.869+0000] {local_task_job_runner.py:123} INFO - ::group::Pre task execution logs
[2024-08-26T10:47:15.890+0000] {taskinstance.py:2603} INFO - Dependencies all met for dep_context=non-requeueable deps ti=<TaskInstance: Spark.spark_clean manual__2024-08-26T10:47:01.375823+00:00 [queued]>
[2024-08-26T10:47:15.899+0000] {taskinstance.py:2603} INFO - Dependencies all met for dep_context=requeueable deps ti=<TaskInstance: Spark.spark_clean manual__2024-08-26T10:47:01.375823+00:00 [queued]>
[2024-08-26T10:47:15.900+0000] {taskinstance.py:2856} INFO - Starting attempt 1 of 1
[2024-08-26T10:47:15.916+0000] {taskinstance.py:2879} INFO - Executing <Task(PythonOperator): spark_clean> on 2024-08-26 10:47:01.375823+00:00
[2024-08-26T10:47:15.923+0000] {warnings.py:112} WARNING - /home/***/.local/lib/python3.12/site-packages/***/task/task_runner/standard_task_runner.py:70: DeprecationWarning: This process (pid=925) is multi-threaded, use of fork() may lead to deadlocks in the child.
  pid = os.fork()

[2024-08-26T10:47:15.925+0000] {standard_task_runner.py:72} INFO - Started process 927 to run task
[2024-08-26T10:47:15.926+0000] {standard_task_runner.py:104} INFO - Running: ['***', 'tasks', 'run', 'Spark', 'spark_clean', 'manual__2024-08-26T10:47:01.375823+00:00', '--job-id', '213', '--raw', '--subdir', 'DAGS_FOLDER/Seoul_spark.py', '--cfg-path', '/tmp/tmp9apxdvwn']
[2024-08-26T10:47:15.928+0000] {standard_task_runner.py:105} INFO - Job 213: Subtask spark_clean
[2024-08-26T10:47:15.991+0000] {task_command.py:467} INFO - Running <TaskInstance: Spark.spark_clean manual__2024-08-26T10:47:01.375823+00:00 [running]> on host 94bdab50145e
[2024-08-26T10:47:16.087+0000] {taskinstance.py:3122} INFO - Exporting env vars: AIRFLOW_CTX_DAG_OWNER='***' AIRFLOW_CTX_DAG_ID='Spark' AIRFLOW_CTX_TASK_ID='spark_clean' AIRFLOW_CTX_EXECUTION_DATE='2024-08-26T10:47:01.375823+00:00' AIRFLOW_CTX_TRY_NUMBER='1' AIRFLOW_CTX_DAG_RUN_ID='manual__2024-08-26T10:47:01.375823+00:00'
[2024-08-26T10:47:16.090+0000] {logging_mixin.py:190} INFO - Task instance is in running state
[2024-08-26T10:47:16.090+0000] {logging_mixin.py:190} INFO -  Previous state of the Task instance: queued
[2024-08-26T10:47:16.091+0000] {logging_mixin.py:190} INFO - Current task name:spark_clean state:running start_date:2024-08-26 10:47:15.891282+00:00
[2024-08-26T10:47:16.092+0000] {logging_mixin.py:190} INFO - Dag name:Spark and current dag run status:running
[2024-08-26T10:47:16.092+0000] {taskinstance.py:731} INFO - ::endgroup::
[2024-08-26T10:47:43.667+0000] {taskinstance.py:3301} ERROR - Task failed with exception
Traceback (most recent call last):
  File "/home/airflow/.local/lib/python3.12/site-packages/airflow/models/taskinstance.py", line 767, in _execute_task
    result = _execute_callable(context=context, **execute_callable_kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/airflow/.local/lib/python3.12/site-packages/airflow/models/taskinstance.py", line 733, in _execute_callable
    return ExecutionCallableRunner(
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/airflow/.local/lib/python3.12/site-packages/airflow/utils/operator_helpers.py", line 252, in run
    return self.func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/airflow/.local/lib/python3.12/site-packages/airflow/models/baseoperator.py", line 406, in wrapper
    return func(self, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/airflow/.local/lib/python3.12/site-packages/airflow/operators/python.py", line 238, in execute
    return_value = self.execute_callable()
                   ^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/airflow/.local/lib/python3.12/site-packages/airflow/operators/python.py", line 256, in execute_callable
    return runner.run(*self.op_args, **self.op_kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/airflow/.local/lib/python3.12/site-packages/airflow/utils/operator_helpers.py", line 252, in run
    return self.func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/airflow/dags/Seoul_spark.py", line 60, in _sqlspark_data_refining
    df_pyspark =  df_pyspark.withColumn("Measurement date", substring(col("Measurement date"), 1, 4))
                                                            ^^^^^^^^^
NameError: name 'substring' is not defined
[2024-08-26T10:47:43.693+0000] {logging_mixin.py:190} INFO - Task instance in failure state
[2024-08-26T10:47:43.695+0000] {logging_mixin.py:190} INFO - Task start:2024-08-26 10:47:15.891282+00:00 end:2024-08-26 10:47:43.692333+00:00 duration:27.801051
[2024-08-26T10:47:43.696+0000] {logging_mixin.py:190} INFO - Task:<Task(PythonOperator): spark_clean> dag:<DAG: Spark> dagrun:<DagRun Spark @ 2024-08-26 10:47:01.375823+00:00: manual__2024-08-26T10:47:01.375823+00:00, state:running, queued_at: 2024-08-26 10:47:01.394752+00:00. externally triggered: True>
[2024-08-26T10:47:43.698+0000] {logging_mixin.py:190} INFO - Failure caused by name 'substring' is not defined
[2024-08-26T10:47:43.700+0000] {taskinstance.py:1225} INFO - Marking task as FAILED. dag_id=Spark, task_id=spark_clean, run_id=manual__2024-08-26T10:47:01.375823+00:00, execution_date=20240826T104701, start_date=20240826T104715, end_date=20240826T104743
[2024-08-26T10:47:43.730+0000] {taskinstance.py:340} INFO - ::group::Post task execution logs
[2024-08-26T10:47:43.732+0000] {standard_task_runner.py:124} ERROR - Failed to execute job 213 for task spark_clean (name 'substring' is not defined; 927)
Traceback (most recent call last):
  File "/home/airflow/.local/lib/python3.12/site-packages/airflow/task/task_runner/standard_task_runner.py", line 117, in _start_by_fork
    ret = args.func(args, dag=self.dag)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/airflow/.local/lib/python3.12/site-packages/airflow/cli/cli_config.py", line 49, in command
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/home/airflow/.local/lib/python3.12/site-packages/airflow/utils/cli.py", line 115, in wrapper
    return f(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^
  File "/home/airflow/.local/lib/python3.12/site-packages/airflow/cli/commands/task_command.py", line 483, in task_run
    task_return_code = _run_task_by_selected_method(args, _dag, ti)
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/airflow/.local/lib/python3.12/site-packages/airflow/cli/commands/task_command.py", line 256, in _run_task_by_selected_method
    return _run_raw_task(args, ti)
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/airflow/.local/lib/python3.12/site-packages/airflow/cli/commands/task_command.py", line 341, in _run_raw_task
    return ti._run_raw_task(
           ^^^^^^^^^^^^^^^^^
  File "/home/airflow/.local/lib/python3.12/site-packages/airflow/utils/session.py", line 97, in wrapper
    return func(*args, session=session, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/airflow/.local/lib/python3.12/site-packages/airflow/models/taskinstance.py", line 2995, in _run_raw_task
    return _run_raw_task(
           ^^^^^^^^^^^^^^
  File "/home/airflow/.local/lib/python3.12/site-packages/airflow/models/taskinstance.py", line 273, in _run_raw_task
    TaskInstance._execute_task_with_callbacks(
  File "/home/airflow/.local/lib/python3.12/site-packages/airflow/models/taskinstance.py", line 3149, in _execute_task_with_callbacks
    result = self._execute_task(context, task_orig)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/airflow/.local/lib/python3.12/site-packages/airflow/models/taskinstance.py", line 3173, in _execute_task
    return _execute_task(self, context, task_orig)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/airflow/.local/lib/python3.12/site-packages/airflow/models/taskinstance.py", line 767, in _execute_task
    result = _execute_callable(context=context, **execute_callable_kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/airflow/.local/lib/python3.12/site-packages/airflow/models/taskinstance.py", line 733, in _execute_callable
    return ExecutionCallableRunner(
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/airflow/.local/lib/python3.12/site-packages/airflow/utils/operator_helpers.py", line 252, in run
    return self.func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/airflow/.local/lib/python3.12/site-packages/airflow/models/baseoperator.py", line 406, in wrapper
    return func(self, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/airflow/.local/lib/python3.12/site-packages/airflow/operators/python.py", line 238, in execute
    return_value = self.execute_callable()
                   ^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/airflow/.local/lib/python3.12/site-packages/airflow/operators/python.py", line 256, in execute_callable
    return runner.run(*self.op_args, **self.op_kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/airflow/.local/lib/python3.12/site-packages/airflow/utils/operator_helpers.py", line 252, in run
    return self.func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/airflow/dags/Seoul_spark.py", line 60, in _sqlspark_data_refining
    df_pyspark =  df_pyspark.withColumn("Measurement date", substring(col("Measurement date"), 1, 4))
                                                            ^^^^^^^^^
NameError: name 'substring' is not defined
[2024-08-26T10:47:43.758+0000] {local_task_job_runner.py:261} INFO - Task exited with return code 1
[2024-08-26T10:47:43.784+0000] {taskinstance.py:3891} INFO - 0 downstream tasks scheduled from follow-on schedule check
[2024-08-26T10:47:43.786+0000] {local_task_job_runner.py:240} INFO - ::endgroup::
